server/src/main/java/com/cloud/network/NetworkServiceImpl.java

```diff
@@ -2267,7 +2267,7 @@ public Pair<List<? extends Network>, Integer> searchForNetworks(ListNetworksCmd
             isRecursive = true;
         }
 
-        Filter searchFilter = new Filter(NetworkVO.class, "id", false, null, null);
+        Filter searchFilter = new Filter(NetworkVO.class, "id", false, cmd.getStartIndex(), cmd.getPageSizeVal());
```

Comment:
if networkfilter is Account, Domain or Shared, it should work

However, if the networkfilter is All or AccountDomain, it might not work if the searchFilter is the same
for example, there are 25 networks for account, 25 networks for domain. if page size is 20
- on first page, it returns 20 networks for account, it is good
- on second page, it should return the last 5 networks for account, and the first 15 networks for domain. But it will return the last 5 networks for account, and the last 5 networks for domain. right ?
I think the `searchFilter` needs to be updated after the each search.

Reply:
That's a very good point. 
The issue here is that we are concatenating results from multiple queries ( lest's say Q1, Q2, Q3). Things **kind of** work for the first page. We can adjust pageSize of a query depending on how many results are already returned for other queries.

But if we want to display the second page, it is hard to say which startIndex should be used for Q2. As it depends on how many results were displayed for Q2 on page 1. Which in turn depends on how many Q1 results were there on Page 1. For example, if page 1 had 15 results for Q1 and 5 results for Q2, then page2 Q2 startIndex should be 6.

So, to fix this I think the filter should always have `startIndex` as `0` and `pageSize` should be `cmd.getPageSizeVal() * (cmd.getStartIndex() + 1)`

 `Filter searchFilter = new Filter(NetworkVO.class, "id", false, 0, (cmd.getStartIndex() + 1) * cmd.getPageSizeVal());`
 
 This way we have info on all previous page results and when pagination is applied at the end of the method, the correct result will be returned.
 
 We are losing some optimization if startIndex is higher but it is still better than reading all rows.

Reply:
Are there not any other API similar to this where pagination is applied where the pattern can be repeated, maybe templates, volumes, offerings?